src/combinatorics/simple_graph/hom.lean

```diff
@@ -0,0 +1,137 @@
+-/
+
+universes u v
+variables {V : Type u} {W : Type v} {G : simple_graph V} {F : simple_graph W}
+
+/-- The subtype of `G.subgraph` comprising those subgraphs with finite vertex sets. -/
+abbreviation simple_graph.fin_subgraph (G : simple_graph V) :=
+  { G' : G.subgraph // G'.verts.finite }
+
+/-- A graph homomorphism from a finite subgraph of G to F. -/
+abbreviation fin_subgraph_hom (G' : G.fin_subgraph) (F : simple_graph W) := G'.val.coe →g F
+infix ` →fg ` : 50 := fin_subgraph_hom
+
+/-- The finite subgraph of G generated by a single vertex. -/
+def vertex_singleton (v : V) : G.fin_subgraph :=
+ ⟨{verts := {v},
+   adj := λ _ _, false,
+   adj_sub := λ _ _, by contradiction,
+   edge_vert := λ _ _, by contradiction},
+  by simp only [set.finite_singleton]⟩
+
+/-- The finite subgraph of G generated by a single edge. -/
+def edge_singleton {u v : V} (e : G.adj u v) : G.fin_subgraph :=
+  ⟨{verts := {u, v},
+    adj := λ x y, (x = u ∧ y = v) ∨ (y = u ∧ x = v),
+    adj_sub := λ x y, by {rintro (⟨rfl, rfl⟩ | ⟨rfl, rfl⟩), exact e, exact e.symm},
+    edge_vert := λ x _, by {rintro (⟨rfl, _⟩ | ⟨_, rfl⟩); simp}},
+  by simp only [set.finite.insert, set.finite_singleton]⟩
+
+/- Lemmas establishing the ordering between edge- and vertex-generated subgraphs. -/
+
+lemma vertex_left_le_edge {u v : V} {e : G.adj u v} : vertex_singleton u ≤ edge_singleton e :=
+begin
+  split,
+  { unfold vertex_singleton, unfold edge_singleton,
+    simp only [subtype.coe_mk, set.singleton_subset_iff, set.mem_insert_iff, eq_self_iff_true,
+               true_or]},
+  { unfold vertex_singleton,
+    simp only [subtype.coe_mk, is_empty.forall_iff, forall_2_true_iff]},
+end
+
+lemma vertex_right_le_edge {u v : V} {e : G.adj u v} : vertex_singleton v ≤ edge_singleton e :=
+begin
+  split,
+  { unfold vertex_singleton, unfold edge_singleton,
+    simp only [subtype.coe_mk, set.singleton_subset_iff, set.mem_insert_iff, set.mem_singleton,
+               or_true]},
+  { unfold vertex_singleton,
+    simp only [subtype.coe_mk, is_empty.forall_iff, forall_2_true_iff]}
+end
+
+/-- Given a homomorphism from a subgraph to `F`, construct its restriction to a sub-subgraph. -/
+def fin_subgraph_hom_restrict {G' G'' : G.fin_subgraph} (h : G'' ≤ G') (f : G' →fg F) : G'' →fg F :=
+begin
+  refine ⟨λ ⟨v, hv⟩, f.to_fun ⟨v, h.1 hv⟩, _⟩,
+  rintros ⟨u, hu⟩ ⟨v, hv⟩ huv,
+  exact f.map_rel' (h.2 huv),
+end
+
+/-- The inverse system of finite homomorphisms. -/
+def fin_subgraph_hom_functor (G : simple_graph V) (F : simple_graph W) :
+  (G.fin_subgraph)ᵒᵖ ⥤ Type (max u v) :=
+{ obj := λ G', G'.unop →fg F,
+  map := λ G' G'' g f, fin_subgraph_hom_restrict (category_theory.le_of_hom g.unop) f, }
+
+/-- If every finite subgraph of a graph `G` has a homomorphism to a finite graph `F`, then there is
+a homomorphism from the whole of `G` to `F`. -/
+lemma simple_graph.exists_hom_of_all_finite_homs [finite W]
+  (h : ∀ (G' : G.fin_subgraph), G' →fg F) : nonempty (G →g F) :=
```

Comment:
Since we're passing actual homomorphisms to this lemma:
```suggestion
  (h : Π (G' : G.fin_subgraph), G' →fg F) : nonempty (G →g F) :=
```
This is just stylistic; it doesn't change anything.